lib/gemstash/s3_storage.rb

```diff
@@ -0,0 +1,290 @@
+# frozen_string_literal: true
+
+require "gemstash"
+require "digest"
+require "fileutils"
+require "yaml"
+require "aws-sdk-s3"
+require "pathname"
+
+module Gemstash
+  # The entry point into the storage engine for storing cached gems, specs, and
+  # private gems using Amazon S3 service.
+  class S3
+    extend Gemstash::Env::Helper
+    include Gemstash::Env::Helper
+    attr_reader :folder, :client, :bucket
+
+    STORAGE_VERSION = 1
+
+    # If the storage engine detects the base cache directory was originally
+    # initialized with a newer version, this error is thrown.
+    class VersionTooNew < StandardError
+      def initialize(folder, version)
+        super("Gemstash storage version #{Gemstash::S3::STORAGE_VERSION} does " \
+               "not support version #{version} found at #{folder}")
+      end
+    end
+
+    def initialize(folder, root: true)
+      @folder = folder
+      check_storage_version if root
+      @client = Aws::S3::Client.new(
+        access_key_id: gemstash_env.config[:aws_access_key_id],
+        secret_access_key: gemstash_env.config[:aws_secret_access_key],
+        region: gemstash_env.config[:region]
+      )
+      @bucket_name = gemstash_env.config[:bucket_name]
+      @object_name = @folder
+    end
+
+    def check_credentials?
+      @client.get_bucket_location(bucket: @bucket_name).location_constraint == gemstash_env.config[:region]
+    end
+
+    def delete_with_prefix(prefix = @folder)
+      Aws::S3::Resource.new(client: @client).bucket(@bucket_name).objects(prefix: prefix).batch_delete!
+    end
+
+    def resource(id)
+      S3Resource.new(@folder, id, @client, @bucket_name)
+    end
+
+    def for(child)
+      S3.new(File.join(@folder, child), root: false)
+    end
+
+    def self.for(name)
+      new(File.join(gemstash_env.config[:s3_path], name))
+    end
+
+    def self.metadata
+      file = gemstash_env.base_file("metadata.yml")
+      unless File.exist?(file)
+        gemstash_env.atomic_write(file) do |f|
+          f.write({ storage_version: Gemstash::S3::STORAGE_VERSION,
+                    gemstash_version: Gemstash::VERSION }.to_yaml)
+        end
+      end
+
+      YAML.load_file(file)
+    end
+
+  private
+
+    def check_storage_version
+      version = Gemstash::S3.metadata[:storage_version]
+      return if version <= Gemstash::S3::STORAGE_VERSION
+
+      raise Gemstash::S3::VersionTooNew.new(@folder, version)
+    end
+  end
+
+  # A resource within the storage engine. The resource may have 1 or more files
+  # associated with it along with a metadata Hash that is stored in a YAML file.
+  class S3Resource
+    include Gemstash::Env::Helper
+    include Gemstash::Logging
+    attr_reader :name, :folder, :client
+    VERSION = 1
+
+    # If the storage engine detects a resource was originally saved from a newer
+    # version, this error is thrown.
+    class VersionTooNew < StandardError
```

Comment:
if this is an erro it shouldbe named ` VersionTooNewError`

Reply:
Fixed in commit #dcd8a50